server/middleware/identityMiddleware.ts

```diff
@@ -1,156 +1,158 @@
-import url from 'url';
-import type express from 'express';
-import fetch from 'node-fetch';
+import type OktaJwtVerifier from '@okta/jwt-verifier';
+import type { NextFunction, Request, Response } from 'express';
+import type { Scopes, VerifiedOAuthCookies } from '@/server/oauth';
 import {
-	getScopeFromRequestPathOrEmptyString,
-	X_GU_ID_FORWARDED_SCOPE,
-} from '../../shared/identity';
+	OAuthAccessTokenCookieName,
+	oauthCookieOptions,
+	OAuthIdTokenCookieName,
+	performAuthorizationCodeFlow,
+	scopes,
+	verifyAccessToken,
+	verifyIdToken,
+} from '@/server/oauth';
 import { requiresSignin } from '../../shared/requiresSignin';
-import { handleAwsRelatedError } from '../awsIntegration';
 import { conf } from '../config';
-import { idapiConfigPromise } from '../idapiConfig';
-import { log } from '../log';
-import type { MockableExpressRequest } from './requestMiddleware';
-import {
-	augmentRedirectURL,
-	signInTokenQueryParameterNames,
-	updateManageUrl,
-} from './requestMiddleware';
 
-interface RedirectResponseBody extends IdentityDetails {
-	signInStatus: string;
-	redirect?: {
-		url: string;
-	};
-}
+declare const CYPRESS: string;
 
-const containsSignInTokenQueryParameters = (
-	req: MockableExpressRequest,
-): boolean =>
-	signInTokenQueryParameterNames.some(
-		(name) => req.query[name] !== undefined,
-	);
+const handleOAuthMiddlewareError = (err: Error, res: Response) => {
+	console.log('OAuth / Middleware / Error', err);
+	res.redirect('/maintenance');
+};
 
-const redirectOrCustomStatusCode = (
-	res: express.Response,
-	redirectURL: string,
-	statusCode?: number,
-) =>
-	statusCode
-		? res.status(statusCode).header('Location', redirectURL).send()
-		: res.redirect(redirectURL);
+// Returns either an object with the verified access and ID tokens, or an empty object.
+export const verifyOAuthCookies = async (
+	req: Request,
+): Promise<VerifiedOAuthCookies> => {
+	const accessTokenCookie = req.signedCookies['GU_ACCESS_TOKEN'];
+	const idTokenCookie = req.signedCookies['GU_ID_TOKEN'];
 
-export const getCookiesOrEmptyString = (req: express.Request) =>
-	req.header('cookie') || '';
+	if (accessTokenCookie && idTokenCookie) {
+		const accessToken = await verifyAccessToken(accessTokenCookie);
+		const idToken = await verifyIdToken(idTokenCookie);
+		if (
+			// check access token is valid
+			accessToken &&
+			// check that the id token is valid
+			idToken &&
+			// check that the access token is not expired
+			!accessToken.isExpired() &&
+			// check that the scopes are all the ones we expect
+			accessToken.claims.scp?.every((scope) =>
+				scopes.includes(scope as Scopes),
```

Comment:
isn't this backwards ?
don't we want to check that every scope on `scopes` is in `accesToken.claims.scp` ?